version-negotiation.md

```diff
@@ -0,0 +1,74 @@
+
+### Kill switch
+
+There needs to be a way for a page to remove any previous origin policies, before the page loads. This is essentially a special case of sync update, which doesn't require loading a new policy since the browser already knows what the null policy looks like.
+
+### Privacy-preserving
+
+Any scenario in which the browser communicates previously-sent information back to the server or website can be (ab)used as a cookie. Previous incarnations of origin policy communicated the currently-downloaded policy version. But even the behaviors controlled by origin policy could be used in this manner.
+
+## Proposed model
+
+Conceptually, origin policies are stored in the HTTP cache, under the URL `$origin/.well-known/origin`. In particular, this means they are double-keyed [in the same way as the HTTP cache](https://github.com/WICG/origin-policy/issues/23): that is, by (origin of document, origin of top-document). This prevents them from being used for cross-site tracking.
+
+(Question: should clearing the HTTP cache also clear any cached origin policy? Probably yes. Or should it be tied to clearing cookies?)
+
+When the browser makes a request to _url_, it:
+
+* Checks if it has something cached, and not expired, for `$that_origin/.well-known/origin-policy`. If so, this is the _candidate origin policy_; otherwise the _candidate origin policy_ is the "none" policy.
+* Makes the request to _url_.
+* The response can contain a header of the form `Origin-Policy: allowed=(policyA polB polC none), preferred=policyZ`. This indicates that policies with identifiers `policyA`, `polB`, `polC`, or `policyZ` are acceptable to the site, or no origin policy at all. Call these the _list of acceptable policies_. If the response contains no such header, then the _list of acceptable policies_ contains just the "none" policy.
+* Checks the _list of acceptable policies_ against the _candidate origin policy_.
+  * If the _candidate origin policy_ is in the _list of acceptable origin policies_, then:
+    * Apply the _candidate origin policy_ and load the response. (This might apply the "none" policy.)
+    * If _candidate origin policy_ is not the preferred origin policy (indicated by the `preferrered=` portion), then the browser sends a low-priority request to `$that_origin/.well-known/origin-policy` to refresh the HTTP cache, but it won't apply for this page load.
+  * Otherwise, if the _list of acceptable origin policies_ only contains the "none" policy but _candidate origin policy_ is not the none policy, then apply the "none" policy anyway, and load the response.
+  * Otherwise, the browser makes a request (on the same connection, if HTTP/2), to `$that_origin/.well-known/origin-policy`. It delays any loading of the response for _url_ until the new policy has been loaded. If the new policy's identifier still doesn't match the _list of acceptable policies_, then the result of the origin navigation request is a network error.
```

Comment:
Loading of _url_ doesn't have to be delayed, but we need to delay processing that would yield an observably different result.